packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,437 @@
+        case 404:
+          return { code: CanonicalCode.NOT_FOUND };
+        case 403:
+          return { code: CanonicalCode.PERMISSION_DENIED };
+        case 401:
+          return { code: CanonicalCode.UNAUTHENTICATED };
+        case 429:
+          return { code: CanonicalCode.RESOURCE_EXHAUSTED };
+        case 501:
+          return { code: CanonicalCode.UNIMPLEMENTED };
+        case 503:
+          return { code: CanonicalCode.UNAVAILABLE };
+        default:
+          return { code: CanonicalCode.UNKNOWN };
+      }
+    }
+  }
+
+  /**
+   * Returns whether the Expect header is on the given options object.
+   * @param options Options for http.request.
+   */
+  static hasExpectHeader(options: RequestOptions | url.URL): boolean {
+    return !!((options as RequestOptions).headers && (options as RequestOptions).headers!.Expect);
+  }
+
+  /**
+   * Check whether the given request match pattern
+   * @param url URL of request
+   * @param request Request to inspect
+   * @param pattern Match pattern
+   */
+  static isSatisfyPattern<T>(url: string, request: T, pattern: IgnoreMatcher<T>): boolean {
+    if (typeof pattern === 'string') {
+      return pattern === url;
+    } else if (pattern instanceof RegExp) {
+      return pattern.test(url);
+    } else if (typeof pattern === 'function') {
+      return pattern(url, request);
+    } else {
+      throw new TypeError('Pattern is in unsupported datatype');
+    }
+  }
+
+  /**
+   * Check whether the given request is ignored by configuration
+   * @param url URL of request
+   * @param request Request to inspect
+   * @param list List of ignore patterns
+   */
+  static isIgnored<T>(url: string, request: T, list?: Array<IgnoreMatcher<T>>): boolean {
+    if (!list) {
+      // No ignored urls - trace everything
+      return false;
+    }
+
+    for (const pattern of list) {
+      if (HttpPlugin.isSatisfyPattern(url, request, pattern)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  static setSpanOnError(span: Span, obj: NodeJS.EventEmitter) {
+    obj.on('error', error => {
+      span
+        .setAttribute(HttpPlugin.ATTRIBUTE_ERROR, true)
+        .setAttribute(HttpPlugin.ATTRIBUTE_HTTP_ERROR_NAME, error.name)
+        .setAttribute(HttpPlugin.ATTRIBUTE_HTTP_ERROR_MESSAGE, error.message);
+      span.setStatus({ code: CanonicalCode.UNKNOWN, message: error.message });
+      span.end();
+    });
+  }
+
+  options!: HttpPluginConfig;
+
+  // TODO: BasePlugin should pass the logger or when we enable the plugin
+  protected readonly _logger!: Logger;
+  protected readonly _tracer!: NodeTracer;
+
+  /** Constructs a new HttpPlugin instance. */
+  constructor(public moduleName: string, public version: string) {
+    super();
+    // TODO: remove this once a logger will be passed
+    this._logger = new NoopLogger();
+  }
+
+  /** Patches HTTP incoming and outcoming request functions. */
+  protected patch() {
```

Comment:
Should this be protected?

Reply:
I would say yes for https / http2 plugins.